### PR TITLE
Query: Fix memory leak by making FeedIterator IDisposable

### LINKED ISSUE #1607: FeedIterator needs to be disposed
Current public surface `FeedIterator` is not disposable, but there are scenarios where disposing it improves memory handling.

This issue is to track adding `IDisposable` to `FeedIterator`.

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description

This PR adds IDisposable to the FeedIterators to fix a memory leak caused by the query logic buffering pages in the background. The FeedIterators have default implementation to avoid breaking any users that are extending the base abstract class.

**NOTE: BREAKING CHANGE**
Any users that have a static analysis tool checking for IDisposable will be broken when upgrade to this version.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Closing issues

Put closes #1607

## Assignee

Please add yourself as the assignee

## Projects

Please add relevant projects so this issue can be properly tracked.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInlineCore.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    using Microsoft.Azure.Cosmos.CosmosElements;
+
+    internal sealed class FeedIteratorInlineCore : FeedIteratorInternal
+    {
+        private readonly FeedIteratorInternal feedIteratorInternal;
+
+        internal FeedIteratorInlineCore(
+            FeedIterator feedIterator)
+        {
+            if (feedIterator is FeedIteratorInternal feedIteratorInternal)
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInlineCore.cs:19]: negate and throw, then get rid of the else:

```
if (!(feedIterator is FeedIteratorInternal feedIteratorInternal))
{
throw
}

this.feedIteratorInternal = feedIteratorInternal;
```

or use the new c# throw in ternary operator:

`this.feedIteratorInternal = feedIterator is FeedIteratorInternal feedIteratorInternal ? feedIteratorInteral : throw ...`

File: Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInlineCore.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            {
+                this.feedIteratorInternal = feedIteratorInternal;
+            }
+            else
+            {
+                throw new ArgumentNullException(nameof(feedIterator));
+            }
+        }
+
+        internal FeedIteratorInlineCore(
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInlineCore.cs:29]: don't have multiple constructors. Create method or make the caller cast.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Query/SanityQueryTests.cs [hunk 3]
```diff
@@ -64,6 +65,128 @@ async Task ImplementationAsync(Container container, IReadOnlyList<CosmosObject>
...
+            QueryOracleUtil util = new QueryOracle2(seed);
+            IEnumerable<string> inputDocuments = util.GetDocuments(numberOfDocuments);
+
+            await this.CreateIngestQueryDeleteAsync(
+                ConnectionModes.Direct,
+                CollectionTypes.MultiPartition,
+                inputDocuments,
+                ImplementationAsync);
+
+            //WeakReference weakReferenceFeedIterator;
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Query/SanityQueryTests.cs:82]: remove commented code?

File: Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInlineCore.cs (hunk 2 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInlineCore.cs:27]: That would cause a massive change and is out of scope o this PR.

**Pull request merged**